src/dachshund/weighted_undirected_graph.rs

```diff
@@ -0,0 +1,182 @@
+    fn get_core_ids(&self) -> &Vec<NodeId> {
+        &self.ids
+    }
+    /// core and non-core IDs are the same for a `SimpleUndirectedGraph`.
+    fn get_non_core_ids(&self) -> Option<&Vec<NodeId>> {
+        Some(&self.ids)
+    }
+    fn get_ids_iter(&self) -> Keys<NodeId, WeightedNode> {
+        self.nodes.keys()
+    }
+    fn get_nodes_iter(&self) -> Values<NodeId, WeightedNode> {
+        self.nodes.values()
+    }
+    fn get_mut_nodes(&mut self) -> &mut FxHashMap<NodeId, WeightedNode> {
+        &mut self.nodes
+    }
+    fn has_node(&self, node_id: NodeId) -> bool {
+        self.nodes.contains_key(&node_id)
+    }
+    fn get_node(&self, node_id: NodeId) -> &WeightedNode {
+        &self.nodes[&node_id]
+    }
+    fn count_edges(&self) -> usize {
+        let mut num_edges: usize = 0;
+        for node in self.nodes.values() {
+            num_edges += node.neighbors.len();
+        }
+        num_edges / 2
+    }
+    fn count_nodes(&self) -> usize {
+        self.nodes.len()
+    }
+    fn create_empty() -> Self {
+        WeightedUndirectedGraph {
+            nodes: FxHashMap::default(),
+            ids: Vec::new(),
+        }
+    }
+}
+impl WeightedUndirectedGraph {
+    pub fn as_input_rows(&self, graph_id: usize) -> String {
+        let mut rows: Vec<String> = Vec::new();
+        for (id, node) in &self.nodes {
+            for e in node.get_edges() {
+                if *id < e.get_neighbor_id() {
+                    rows.push(format!(
+                        "{}\t{}\t{}\t{}",
+                        graph_id,
+                        id.value(),
+                        e.get_neighbor_id().value(),
+                        e.weight
+                    ));
+                }
+            }
+        }
+        rows.join("\n")
+    }
+    pub fn get_node_degree(&self, id: NodeId) -> usize {
+        self.nodes[&id].degree()
+    }
+    pub fn get_node_weight(&self, id: NodeId) -> f64 {
+        self.nodes[&id].weight()
+    }
+}
+impl UndirectedGraph for WeightedUndirectedGraph {}
+
+impl ConnectedComponents for WeightedUndirectedGraph {}
+impl ConnectedComponentsUndirected for WeightedUndirectedGraph {}
+impl Coreness for WeightedUndirectedGraph {}
+impl FractionalCoreness for WeightedUndirectedGraph {
+    fn get_fractional_coreness_values(&self) -> HashMap<NodeId, f64> {
+        // Start by making a priority queue of each node with its weight as a priority.
+        // Use PriorityQueue instead of BinaryHeap because the workload uses change priority.
+
+        // [TODO Fix trait bounds and move this to coreness.rs.]
+        // [TODO:Perf] Switch to hashbrown. Benchmark performance.
+        let mut pq = PriorityQueue::with_capacity(self.nodes.len());
+
+        for node in self.get_nodes_iter() {
+            pq.push(node.get_id(), Reverse(NotNan::new(node.weight()).unwrap()));
+        }
+        let mut coreness: HashMap<NodeId, f64> = HashMap::new();
+
+        while !pq.is_empty() {
```

Comment:
I flipped into "expect documentation" mode thanks to the comment above that begins with "Start by making a priority queue", so around this point I start feeling bereft & looking for another helpful comment explaining what's about to happen with coreness shells :)

Reply:
Agreed that the documentation here totally sucks. I added some more comments and will add that to this PR before merging.